Dynamic GRU quantization support

raghuramank100 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49448 Dynamic GRU quantization support**
Enables support for nn.quantized.GRU module for both FBGEMM and QNNPACK.
Fixes disabled tests for RNNCells and LSTM modules.
Ensures that reduce range is turned on by default for FBGEMM for dynamic quantization and is ignored for QNNPACK

Differential Revision: [D25579815](https://our.internmc.facebook.com/intern/diff/D25579815/)

raghuramank10000 committed:
Dynamic GRU quantization support



Differential Revision: [D25579815](https://our.internmc.facebook.com/intern/diff/D25579815/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic GRU quantization support"


Differential Revision: [D25579815](https://our.internmc.facebook.com/intern/diff/D25579815/)

[ghstack-poisoned]

vkuzo commented on test/quantization/test_quantize.py:
```diff
@@ -940,40 +940,64 @@ def checkQuantized(model):
-        # We cannot trace with input dtype being a packed sequence
-        self._checkScriptable(model_with_packed_input, scripted, [[packed_input]], True)
+
+        for module_type in ['LSTM', 'GRU']:
```
should we have a separate test case for each?

raghuramank100 replied:
Sure, can refactor the tests with some common code

vkuzo commented on test/quantization/test_quantize.py:
```diff
@@ -940,40 +940,64 @@ def checkQuantized(model):
+            checkQuantized(model_quantized, module_type)
+            self.checkScriptable(model_quantized, [[x]], check_save_load=True)
+
+            class ScriptWrapperPacked(torch.nn.Module):
```
`ScriptWrapperPackedLSTM`?

raghuramank100 replied:
Sure, will change that

vkuzo commented on test/quantization/test_quantized_op.py:
```diff
@@ -2532,6 +2532,7 @@ def test_qlstmGRU(self, num_batches, input_size, hidden_size,
                 if torch.backends.quantized.engine == 'qnnpack' and dtype == torch.float16:
                     continue
 
+                print(rnn_type, dtype, num_directions, torch.backends.quantized.engine)
```
remove?

raghuramank100 replied:
Yes, currently the test is failing on certain machines, not able to repro locally, will remove once I figure out the cause :-)

vkuzo commented on test/quantization/test_quantized_module.py:
```diff
@@ -1016,6 +1016,55 @@ def test_lstm_api(self, dtype, bidirectional):
+
+
+            y, h = cell_dq(x, h)
+            self.assertEqual(result[0], y, msg="GRU module API failed")
+            self.assertEqual(result[1], h, msg="GRU module API failed")
```
I'm trusting that this would flag any numerical issues, didn't review the formulas line by line (lmk if that is needed for this review)

raghuramank100 replied:
No, we should be ok with that

jerryzh168 commented:
do you plan to add this to fx graph mode as well?

raghuramank10000 committed:
Update on "Dynamic GRU quantization support"


Differential Revision: [D25579815](https://our.internmc.facebook.com/intern/diff/D25579815/)

[ghstack-poisoned]

raghuramank100 commented:
> do you plan to add this to fx graph mode as well?

Yes, will do in a follow up PR

raghuramank10000 committed:
Update on "Dynamic GRU quantization support"


Differential Revision: [D25579815](https://our.internmc.facebook.com/intern/diff/D25579815/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic GRU quantization support"


Differential Revision: [D25579815](https://our.internmc.facebook.com/intern/diff/D25579815/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic GRU quantization support"


Differential Revision: [D25579815](https://our.internmc.facebook.com/intern/diff/D25579815/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic GRU quantization support"

Enables support for nn.quantized.GRU module for both FBGEMM and QNNPACK.
Fixes disabled tests for RNNCells and LSTM modules.
Ensures that reduce range is turned on by default for FBGEMM for dynamic quantization and is ignored for QNNPACK

Differential Revision: [D25579815](https://our.internmc.facebook.com/intern/diff/D25579815/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic GRU quantization support"

Enables support for nn.quantized.GRU module for both FBGEMM and QNNPACK.
Fixes disabled tests for RNNCells and LSTM modules.
Ensures that reduce range is turned on by default for FBGEMM for dynamic quantization and is ignored for QNNPACK

Differential Revision: [D25579815](https://our.internmc.facebook.com/intern/diff/D25579815/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic GRU quantization support"

Enables support for nn.quantized.GRU module for both FBGEMM and QNNPACK.
Fixes disabled tests for RNNCells and LSTM modules.
Ensures that reduce range is turned on by default for FBGEMM for dynamic quantization and is ignored for QNNPACK

Differential Revision: [D25579815](https://our.internmc.facebook.com/intern/diff/D25579815/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Dynamic GRU quantization support"

Enables support for nn.quantized.GRU module for both FBGEMM and QNNPACK.
Fixes disabled tests for RNNCells and LSTM modules.
Ensures that reduce range is turned on by default for FBGEMM for dynamic quantization and is ignored for QNNPACK

Differential Revision: [D25579815](https://our.internmc.facebook.com/intern/diff/D25579815/)

[ghstack-poisoned]
